edxval/models.py

```diff
@@ -335,45 +336,116 @@ def image_url(self):
+    """
+    def __init__(self, *args, **kwargs):
+        kwargs.update(dict(
+            upload_to=video_transcript_path,
+            storage=get_video_transcript_storage(),
+            max_length=255,  # enoungh for uuid
+            blank=True,
+            null=True
+        ))
+        super(CustomizableFileField, self).__init__(*args, **kwargs)
 
-    Attributes:
-        video: the video that the subtitles are for
-        fmt: the format of the subttitles file
+    def deconstruct(self):
+        """
+        Override base class method.
+        """
+        name, path, args, kwargs = super(CustomizableFileField, self).deconstruct()
+        del kwargs['upload_to']
+        del kwargs['storage']
+        del kwargs['max_length']
+        return name, path, args, kwargs
+
+
+class VideoTranscript(TimeStampedModel):
     """
-    created = models.DateTimeField(auto_now_add=True)
-    modified = models.DateTimeField(auto_now=True)
-    video = models.ForeignKey(Video, related_name="subtitles")
-    fmt = models.CharField(max_length=20, db_index=True, choices=SUBTITLE_FORMATS)
-    language = models.CharField(max_length=8, db_index=True)
-    content = models.TextField(default='')
+    Transcript for a video
+    """
+    video_id = models.CharField(max_length=255, help_text='It can be an edx_video_id or an external video id')
+    transcript = CustomizableFileField()
+    language_code = models.CharField(max_length=8, db_index=True)
+    provider = models.CharField(
+        max_length=30,
+        choices=TranscriptProviderType.CHOICES,
+        default=TranscriptProviderType.CUSTOM,
+    )
+    file_format = models.CharField(max_length=20, db_index=True, choices=TranscriptFormat.CHOICES)
 
-    def __str__(self):
-        return '%s Subtitle for %s' % (self.language, self.video)
+    class Meta:
+        unique_together = ('video_id', 'language_code')
 
-    def get_absolute_url(self):
+    @classmethod
+    def create_or_update(cls, video_id, language_code, file_name, file_format, provider, file_data=None):
         """
-        Returns the full url link to the edx_video_id
+        Create or update Transcript object.
+
+        Arguments:
+            video_id (str): unique id for a video
+            language_code (str): language code
+            file_name (str): File name of the image
+            file_format (str): Format of transcript
+            provider (str): Transcript provider
+            file_data (InMemoryUploadedFile): File data to be saved
+
+        Returns:
+            Returns a tuple of (video_transcript, created).
         """
-        return reverse('subtitle-content', args=[self.video.edx_video_id, self.language])
+        video_transcript, created = cls.objects.get_or_create(video_id=video_id, language_code=language_code)
+        video_transcript.transcript.name = file_name
+        video_transcript.file_format = file_format
+        video_transcript.provider = provider
+
+        if file_data:
```

Comment:
I think we should also delete the existing file if any. Like if someone uploads transcript for existing video & language, which is already on s3 then we should delete it before saving a new one because that would be overridden anyway.